File: front/monte-carlo-moments.mjs
export const MONTE_CARLO_METRIC_DEFINITIONS = Object.freeze([
  { metricId: "mission_success_rate", label: "任务可靠度", unit: "比例", varianceUnit: "比例²", valueFormat: "ratio" },
  { metricId: "operational_availability", label: "使用可用度(A)", unit: "比例", varianceUnit: "比例²", valueFormat: "ratio" },
  { metricId: "spare_fill_rate", label: "实际即时满足率", unit: "比例", varianceUnit: "比例²", valueFormat: "ratio", numeratorField: "spare_immediately_filled_total", denominatorField: "spare_demand_total" },
  { metricId: "spare_utilization", label: "实际备件利用率", unit: "比例", varianceUnit: "比例²", valueFormat: "ratio", numeratorField: "spare_consumed_total", denominatorField: "spare_carried_total" },
  { metricId: "ready_rate", label: "战备完好率", unit: "比例", varianceUnit: "比例²", valueFormat: "ratio" },
  { metricId: "sortie_rate", label: "出动架次率", unit: "架次/机/天", varianceUnit: "(架次/机/天)²", valueFormat: "number" },
  { metricId: "mean_transport_delay", label: "平均备件延误时间", unit: "小时", varianceUnit: "小时²", valueFormat: "number" },
  { metricId: "repair_backlog", label: "维修积压", unit: "项", varianceUnit: "项²", valueFormat: "number" }
]);

function isFiniteNumber(value) {
  return typeof value === "number" && Number.isFinite(value);
}

function count(value, fallback = 0) {
  return isFiniteNumber(value) ? Math.max(0, Math.trunc(value)) : fallback;
}

function finiteMean(values) {
  if (!values.length) return null;
  const mean = values.reduce((sum, value) => sum + value / values.length, 0);
  return isFiniteNumber(mean) ? mean : null;
}

function finiteSampleVariance(values, mean) {
  if (values.length < 2 || !isFiniteNumber(mean)) return { value: null, invalidReason: null };
  const directVariance = values.reduce(
    (sum, value) => sum + ((value - mean) ** 2),
    0
  ) / (values.length - 1);
  if (isFiniteNumber(directVariance)) return { value: directVariance, invalidReason: null };
  const scale = Math.max(...values.map((value) => Math.abs(value)));
  if (scale === 0) return { value: 0, invalidReason: null };
  const scaledMean = mean / scale;
  const scaledVariance = values.reduce(
    (sum, value) => sum + ((value / scale - scaledMean) ** 2),
    0
  ) / (values.length - 1);
  const standardDeviation = scale * Math.sqrt(scaledVariance);
  if (!isFiniteNumber(standardDeviation) || standardDeviation > Math.sqrt(Number.MAX_VALUE)) {
    return { value: null, invalidReason: "sample_variance_not_finite" };
  }
  const variance = standardDeviation * standardDeviation;
  return isFiniteNumber(variance)
    ? { value: variance, invalidReason: null }
    : { value: null, invalidReason: "sample_variance_not_finite" };
}

export function buildMonteCarloMetricMoments(samples, counts = {}) {
  const safeSamples = Array.isArray(samples) ? samples : [];
  const metrics = MONTE_CARLO_METRIC_DEFINITIONS.map((definition) => {
    const ratioPairs = definition.numeratorField
      ? safeSamples.flatMap((sample) => {
        const metrics = sample?.metrics || sample?.final || {};
        const numerator = metrics[definition.numeratorField];
        const denominator = metrics[definition.denominatorField];
        return isFiniteNumber(numerator) && numerator >= 0 && isFiniteNumber(denominator) && denominator > 0
          ? [[numerator, denominator]]
          : [];
      })
      : [];
    const zeroDenominatorSampleCount = definition.numeratorField
      ? safeSamples.filter((sample) => {
        const metrics = sample?.metrics || sample?.final || {};
        const numerator = metrics[definition.numeratorField];
        const denominator = metrics[definition.denominatorField];
        return isFiniteNumber(numerator) && numerator >= 0 && denominator === 0;
      }).length
      : 0;
    const values = definition.numeratorField
      ? ratioPairs.map(([numerator, denominator]) => numerator / denominator)
      : safeSamples
        .map((sample) => sample?.metrics?.[definition.metricId] ?? sample?.final?.[definition.metricId])
        .filter(isFiniteNumber);
    const validSampleCount = values.length;
    const arithmeticMean = finiteMean(values);
    const numeratorTotal = ratioPairs.length ? ratioPairs.reduce((total, [numerator]) => total + numerator, 0) : null;
    const denominatorTotal = ratioPairs.length ? ratioPairs.reduce((total, [, denominator]) => total + denominator, 0) : null;
    const mean = arithmeticMean;
    const overallRatio = definition.numeratorField && denominatorTotal > 0 ? numeratorTotal / denominatorTotal : null;
    const varianceResult = finiteSampleVariance(values, arithmeticMean);
    const invalidReason = validSampleCount > 0 && mean === null
      ? "mean_not_finite"
      : varianceResult.invalidReason;
    return {
      ...definition,
      mean,
      sampleVariance: varianceResult.value,
      validSampleCount,
      invalidReason: definition.numeratorField && validSampleCount === 0
        ? zeroDenominatorSampleCount ? "zero_denominator" : "data_unavailable"
        : invalidReason,
      meanAggregationMethod: "arithmetic_mean",
      overallAggregationMethod: definition.numeratorField ? "ratio_of_totals" : null,
      overallStatus: overallRatio !== null ? "available" : zeroDenominatorSampleCount ? "zero_denominator" : "data_unavailable",
      numeratorTotal,
      denominatorTotal,
      overallRatio
    };
  });
  const successfulSampleCount = count(counts.successfulSampleCount, safeSamples.length);
  const failedSampleCount = count(counts.failedSampleCount);
  return {
    schemaVersion: "monte-carlo-metric-moments-v0",
    varianceMethod: "unbiased_sample_variance",
    varianceDenominator: "n-1",
    totalSampleCount: count(counts.totalSampleCount, successfulSampleCount + failedSampleCount),
    successfulSampleCount,
    failedSampleCount,
    metrics
  };
}

export function normalizeMonteCarloMetricMoments(payload, samples = [], counts = {}) {
  if (!payload || !Array.isArray(payload.metrics)) {
    return buildMonteCarloMetricMoments(samples, counts);
  }
  const byId = new Map(payload.metrics
    .filter((metric) => metric && typeof metric.metric_id === "string")
    .map((metric) => [metric.metric_id, metric]));
  const metrics = MONTE_CARLO_METRIC_DEFINITIONS.map((definition) => {
    const source = byId.get(definition.metricId) || {};
    const validSampleCount = count(source.valid_sample_count);
    const mean = validSampleCount > 0 && isFiniteNumber(source.mean) ? source.mean : null;
    const sampleVariance = validSampleCount >= 2
      && isFiniteNumber(source.sample_variance)
      && source.sample_variance >= 0
      ? source.sample_variance
      : null;
    const invalidReason = typeof source.invalid_reason === "string" && source.invalid_reason
      ? source.invalid_reason
      : validSampleCount > 0 && mean === null
        ? "mean_not_finite"
        : validSampleCount >= 2 && sampleVariance === null
          ? "sample_variance_not_finite"
          : null;
    return {
      ...definition,
      mean,
      sampleVariance,
      validSampleCount,
      invalidReason,
      meanAggregationMethod: source.mean_aggregation_method || "arithmetic_mean",
      overallAggregationMethod: source.overall_aggregation_method
        || (source.aggregation_method === "ratio_of_totals" ? "ratio_of_totals" : definition.numeratorField ? "ratio_of_totals" : null),
      overallStatus: source.overall_status || (isFiniteNumber(source.overall_ratio) ? "available" : "data_unavailable"),
      numeratorTotal: isFiniteNumber(source.numerator_total) ? source.numerator_total : null,
      denominatorTotal: isFiniteNumber(source.denominator_total) ? source.denominator_total : null,
      overallRatio: isFiniteNumber(source.overall_ratio) ? source.overall_ratio : null
    };
  });
  const successfulSampleCount = count(
    payload.successful_sample_count,
    count(counts.successfulSampleCount, Array.isArray(samples) ? samples.length : 0)
  );
  const failedSampleCount = count(payload.failed_sample_count, count(counts.failedSampleCount));
  return {
    schemaVersion: payload.schema_version || "monte-carlo-metric-moments-v0",
    varianceMethod: "unbiased_sample_variance",
    varianceDenominator: "n-1",
    totalSampleCount: count(
      payload.total_sample_count,
      count(counts.totalSampleCount, successfulSampleCount + failedSampleCount)
    ),
    successfulSampleCount,
    failedSampleCount,
    metrics
  };
}

export function formatMonteCarloMoment(value, _valueFormat, { variance = false } = {}) {
  if (!isFiniteNumber(value)) return variance ? "不可计算" : "--";
  if (variance) return value.toFixed(4);
  return value.toFixed(2);
}

File: src/spare_mvp_contract/monte_carlo_moments.py
"""Canonical Monte Carlo moments for product-facing aircraft support metrics."""

from __future__ import annotations

import math
import sys
from typing import Any


MONTE_CARLO_METRIC_DEFINITIONS: tuple[dict[str, str], ...] = (
    {
        "metric_id": "mission_success_rate",
        "label": "任务可靠度",
        "unit": "比例",
        "variance_unit": "比例²",
        "value_format": "ratio",
    },
    {
        "metric_id": "operational_availability",
        "label": "使用可用度(A)",
        "unit": "比例",
        "variance_unit": "比例²",
        "value_format": "ratio",
    },
    {
        "metric_id": "spare_fill_rate",
        "label": "备件满足率",
        "unit": "比例",
        "variance_unit": "比例²",
        "value_format": "ratio",
    },
    {
        "metric_id": "spare_utilization",
        "label": "备件利用率",
        "unit": "比例",
        "variance_unit": "比例²",
        "value_format": "ratio",
    },
    {
        "metric_id": "ready_rate",
        "label": "战备完好率",
        "unit": "比例",
        "variance_unit": "比例²",
        "value_format": "ratio",
    },
    {
        "metric_id": "sortie_rate",
        "label": "出动架次率",
        "unit": "架次/机/天",
        "variance_unit": "(架次/机/天)²",
        "value_format": "number",
    },
    {
        "metric_id": "mean_transport_delay",
        "label": "平均备件延误时间",
        "unit": "小时",
        "variance_unit": "小时²",
        "value_format": "number",
    },
    {
        "metric_id": "repair_backlog",
        "label": "维修积压",
        "unit": "项",
        "variance_unit": "项²",
        "value_format": "number",
    },
)

RATIO_OF_TOTALS_METRICS: dict[str, tuple[str, str]] = {
    "spare_fill_rate": ("spare_immediately_filled_total", "spare_demand_total"),
    "spare_utilization": ("spare_consumed_total", "spare_carried_total"),
}


def is_finite_json_number(value: Any) -> bool:
    """Return true only for finite native JSON numbers, never booleans or strings."""

    if not isinstance(value, (int, float)) or isinstance(value, bool):
        return False
    try:
        return math.isfinite(float(value))
    except OverflowError:
        return False


def finite_mean(values: list[float]) -> float | None:
    """Return a finite arithmetic mean without overflowing the intermediate sum."""

    if not values:
        return None
    count = len(values)
    try:
        mean = math.fsum(value / count for value in values)
    except (OverflowError, ValueError):
        return None
    return mean if math.isfinite(mean) else None


def _finite_sample_variance(values: list[float], mean: float | None) -> tuple[float | None, str | None]:
    if len(values) < 2 or mean is None:
        return None, None
    try:
        direct_variance = math.fsum((value - mean) ** 2 for value in values) / (len(values) - 1)
        if math.isfinite(direct_variance):
            return direct_variance, None
    except (OverflowError, ValueError):
        pass
    scale = max(abs(value) for value in values)
    if scale == 0:
        return 0.0, None
    try:
        scaled_mean = mean / scale
        scaled_sum_squares = math.fsum((value / scale - scaled_mean) ** 2 for value in values)
        scaled_variance = scaled_sum_squares / (len(values) - 1)
        standard_deviation = scale * math.sqrt(scaled_variance)
        if not math.isfinite(standard_deviation) or standard_deviation > math.sqrt(sys.float_info.max):
            return None, "sample_variance_not_finite"
        variance = standard_deviation * standard_deviation
    except (OverflowError, ValueError):
        return None, "sample_variance_not_finite"
    if not math.isfinite(variance):
        return None, "sample_variance_not_finite"
    return variance, None


def build_monte_carlo_metric_moments(
    samples: list[dict[str, Any]],
    *,
    total_sample_count: int,
    failed_sample_count: int,
) -> dict[str, Any]:
    """Build ordered per-metric mean and unbiased sample variance statistics."""

    successful_sample_count = len(samples)
    metrics = []
    for definition in MONTE_CARLO_METRIC_DEFINITIONS:
        metric_id = definition["metric_id"]
        ratio_fields = RATIO_OF_TOTALS_METRICS.get(metric_id)
        ratio_pairs: list[tuple[float, float]] = []
        unavailable_sample_count = 0
        zero_denominator_sample_count = 0
        if ratio_fields:
            numerator_field, denominator_field = ratio_fields
            for sample in samples:
                sample_metrics = sample.get("metrics") if isinstance(sample, dict) else None
                if not isinstance(sample_metrics, dict):
                    unavailable_sample_count += 1
                    continue
                numerator = sample_metrics.get(numerator_field)
                denominator = sample_metrics.get(denominator_field)
                if not is_finite_json_number(numerator) or not is_finite_json_number(denominator):
                    unavailable_sample_count += 1
                    continue
                numerator_value = float(numerator)
                denominator_value = float(denominator)
                if numerator_value < 0 or denominator_value < 0:
                    unavailable_sample_count += 1
                elif denominator_value == 0:
                    zero_denominator_sample_count += 1
                else:
                    ratio_pairs.append((numerator_value, denominator_value))
            values = [numerator / denominator for numerator, denominator in ratio_pairs]
        else:
            values = [
                float(value)
                for sample in samples
                if isinstance(sample, dict)
                for sample_metrics in [sample.get("metrics")]
                if isinstance(sample_metrics, dict)
                for value in [sample_metrics.get(metric_id)]
                if is_finite_json_number(value)
            ]
        valid_sample_count = len(values)
        arithmetic_mean = finite_mean(values)
        numerator_total = math.fsum(pair[0] for pair in ratio_pairs) if ratio_pairs else None
        denominator_total = math.fsum(pair[1] for pair in ratio_pairs) if ratio_pairs else None
        overall_ratio = (
            numerator_total / denominator_total
            if ratio_fields and denominator_total is not None and denominator_total > 0
            else None
        )
        mean = arithmetic_mean
        sample_variance, invalid_reason = _finite_sample_variance(values, arithmetic_mean)
        if valid_sample_count and mean is None:
            invalid_reason = "mean_not_finite"
        if ratio_fields and not valid_sample_count:
            invalid_reason = "zero_denominator" if zero_denominator_sample_count else "data_unavailable"
        metrics.append(
            {
                **definition,
                "mean": mean,
                "sample_variance": sample_variance,
                "valid_sample_count": valid_sample_count,
                "invalid_reason": invalid_reason,
                "mean_aggregation_method": "arithmetic_mean",
                **({
                    "numerator_field": ratio_fields[0],
                    "denominator_field": ratio_fields[1],
                    "numerator_total": numerator_total,
                    "denominator_total": denominator_total,
                    "overall_ratio": overall_ratio,
                    "overall_aggregation_method": "ratio_of_totals",
                    "overall_status": (
                        "available"
                        if overall_ratio is not None
                        else "zero_denominator" if zero_denominator_sample_count else "data_unavailable"
                    ),
                    "zero_denominator_sample_count": zero_denominator_sample_count,
                    "unavailable_sample_count": unavailable_sample_count,
                } if ratio_fields else {}),
            }
        )

    return {
        "schema_version": "monte-carlo-metric-moments-v0",
        "variance_method": "unbiased_sample_variance",
        "variance_denominator": "n-1",
        "total_sample_count": max(0, int(total_sample_count)),
        "successful_sample_count": successful_sample_count,
        "failed_sample_count": max(0, int(failed_sample_count)),
        "metrics": metrics,
    }
